legacy/bmt.go

```diff
@@ -0,0 +1,615 @@
+	for l := 0; len(anc) > 0; l++ {
+		var nodes []*node
+		hash := []string{""}
+		for i, n := range anc {
+			hash = append(hash, fmt.Sprintf("%v|%v", hashstr(n.left), hashstr(n.right)))
+			if i%2 == 0 && n.parent != nil {
+				nodes = append(nodes, n.parent)
+			}
+		}
+		hash = append(hash, "")
+		hashes = append(hashes, hash)
+		anc = nodes
+	}
+	hashes = append(hashes, []string{"", fmt.Sprintf("%v", hashstr(hash)), ""})
+	total := 60
+	del := "                             "
+	var rows []string
+	for i := len(hashes) - 1; i >= 0; i-- {
+		var textlen int
+		hash := hashes[i]
+		for _, s := range hash {
+			textlen += len(s)
+		}
+		if total < textlen {
+			total = textlen + len(hash)
+		}
+		delsize := (total - textlen) / (len(hash) - 1)
+		if delsize > len(del) {
+			delsize = len(del)
+		}
+		row := fmt.Sprintf("%v: %v", len(hashes)-i-1, strings.Join(hash, del[:delsize]))
+		rows = append(rows, row)
+
+	}
+	rows = append(rows, strings.Join(left, "  "))
+	rows = append(rows, strings.Join(right, "  "))
+	return strings.Join(rows, "\n") + "\n"
+}
+
+// newTree initialises a tree by building up the nodes of a BMT
+// - segment size is stipulated to be the size of the hash
+func newTree(segmentSize, depth int, hashfunc func() hash.Hash) *tree {
+	n := newNode(0, nil, hashfunc())
+	prevlevel := []*node{n}
+	// iterate over levels and creates 2^(depth-level) nodes
+	// the 0 level is on double segment sections so we start at depth - 2 since
+	count := 2
+	for level := depth - 2; level >= 0; level-- {
+		nodes := make([]*node, count)
+		for i := 0; i < count; i++ {
+			parent := prevlevel[i/2]
+			var hasher hash.Hash
+			if level == 0 {
+				hasher = hashfunc()
+			}
+			nodes[i] = newNode(i, parent, hasher)
+		}
+		prevlevel = nodes
+		count *= 2
+	}
+	// the datanode level is the nodes on the last level
+	return &tree{
+		leaves:  prevlevel,
+		result:  make(chan []byte),
+		section: make([]byte, 2*segmentSize),
+	}
+}
+
+// Count returns the maximum amount of bytes that will be processed by this hasher implementation
+//
+// Implements bmt.Hash
+func (h *Hasher) Capacity() int {
+	return h.pool.Size
+}
+
+// writeSection is not in use for this implementation
+//
+// Implements bmt.Hash
+func (h *Hasher) WriteSection(idx int, data []byte) error {
+	return errors.New("This hasher only implements sequential writes. Please use Write() instead")
+}
+
+// SetSpan sets the span length value prefix for the current hash operation
+//
+// Implements bmt.Hash
+func (h *Hasher) SetSpan(length int64) error {
+	span := LengthToSpan(length)
+	h.getTree().span = span
+	return nil
+}
+
+// Size implements hash.Hash
+func (h *Hasher) Size() int {
+	return h.pool.SegmentSize
+}
+
+// BlockSize implements hash.Hash and file.SectionWriter
+func (h *Hasher) BlockSize() int {
+	return 2 * h.pool.SegmentSize
+}
+
+// Sum returns the BMT root hash of the buffer
+// using Sum presupposes sequential synchronous writes (io.Writer interface)
+//
+// Implements hash.Hash
```

Comment:
these comments could probably be removed and replaced by a nil interface check

Reply:
Shouldn't we document the fact that it implements an interface? That seems natural to me?

Reply:
a nil interface check would do; self-documenting code